File: scratch/compact-guards-integration-20260916/compact/src/Test/StateMonadFFI.go
package Test_StateMonadFFI


type stateResult struct {
	val   interface{}
	state int
}

type State func(int) stateResult

func runState(s State, init int) stateResult {
	return s(init)
}

func bindState(s State, g func(interface{}) State) State {
	return func(state int) stateResult {
		r1 := s(state)
		gPrime := g(r1.val)
		return gPrime(r1.state)
	}
}

func pureState(a interface{}) State {
	return func(s int) stateResult {
		return stateResult{val: a, state: s}
	}
}

func get() State {
	return func(s int) stateResult {
		return stateResult{val: s, state: s}
	}
}

func put(s int) State {
	return func(_ int) stateResult {
		return stateResult{val: nil, state: s}
	}
}

func modify(f func(int) int) State {
	return bindState(get(), func(s interface{}) State {
		return put(f(s.(int)))
	})
}

func chainModifications(n int) State {
	if n == 0 {
		return pureState(nil)
	}
	return bindState(modify(func(x int) int { return x + 1 }), func(_ interface{}) State {
		return chainModifications(n - 1)
	})
}

func runManyTimes_StateMonad(n int, depth int, acc int) int {
	if n == 0 {
		return acc
	}
	return runManyTimes_StateMonad(n-1, depth, acc+runState(chainModifications(depth), 0).state)
}

func RunStateMonadFFI(limit int) int {
	// The shared FFI wrapper supplies depth 60; the PureScript test uses 20 repetitions.
	return runManyTimes_StateMonad(20, limit, 0)
}

File: scratch/record-scalarization-integration-20260916/step1/src/Test/StateMonadFFI.java
    private record StateResult<A>(A value, int state) {}

    @FunctionalInterface
    private interface State<A> {
        StateResult<A> run(int initial);
    }

    private static <A, B> State<B> bindState(State<A> action, java.util.function.Function<A, State<B>> next) {
        return initial -> {
            StateResult<A> first = action.run(initial);
            return next.apply(first.value()).run(first.state());
        };
    }

    private static <A> State<A> pureState(A value) {
        return initial -> new StateResult<>(value, initial);
    }

    private static State<Integer> getState() {
        return initial -> new StateResult<>(initial, initial);
    }

    private static State<Void> putState(int value) {
        return ignored -> new StateResult<>(null, value);
    }

    private static State<Void> modifyState(java.util.function.IntUnaryOperator update) {
        return bindState(getState(), current -> putState(update.applyAsInt(current)));
    }

    private static State<Void> chainModifications(int depth) {
        if (depth == 0) return pureState(null);
        return bindState(modifyState(value -> value + 1), ignored -> chainModifications(depth - 1));
    }

    public static final java.util.function.Function<Object, Object> runStateMonadFFI = Bench.nativeBenchmark(input -> {
        // The shared FFI wrapper supplies depth 60; the PureScript test supplies 20 repetitions.
        int depth = (Integer) input;
        int result = 0;
        for (int i = 0; i < 20; i++) {
            result += chainModifications(depth).run(0).state();
        }
        return result;
    });
